porch/api/porchconfig/v1alpha1/types_functions.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+//+kubebuilder:resource:path=functions,singular=function
+
+// Function represents a kpt function discovered in a repository
+// Function resources are created automatically by discovery in a registered Repository.
+// Function resource names will be computed as <Repository Name>:<Function Name>
+// to ensure uniqueness of names, and will follow formatting of
+// [DNS Subdomain Names](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names).
+type Function struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   FunctionSpec   `json:"spec,omitempty"`
+	Status FunctionStatus `json:"status,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+
+// FunctionList
+type FunctionList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []Function `json:"items"`
+}
+
+type FunctionType string
+
+const (
+	FunctionTypeValidator FunctionType = "validator"
+	FunctionTypeMutator   FunctionType = "mutator"
+)
+
+// FunctionSpec defines the desired state of a Function
+type FunctionSpec struct {
+	// Image specifies the function image, such as 'gcr.io/kpt-fn/gatekeeper:v0.2'.
+	Image string `json:"image"`
```

Comment:
This is more of a question:
Since this will act as function discovery API, it need to surface all the versions, right ?

For use-cases of resolving the fully qualified domain or migrating to latest version, being able to discover all the versions is important.

Reply:
Well I think we want to think more about this.  I quite like the "Catalog" approach from kustomize.  I think most end-users just want to run the "foo" function.  I think we might need functionality for the platform-team to (safely) update the "foo" function to a newer version.

In short, function discovery is pretty slow and not necessarily the most useful thing we could be doing.

But I think that is orthogonal to this PR, so I'm going to merge and we should discuss in next f2f meeting!